### Leveraging transcriptomic profiles and deep learning to detect homologous recombination deficiency in breast cancer with ***softHRD***

## Intro:
  The homologous recombination (HR) pathway is the canonical repair mechanism by which cells repair DNA double-strand breaks. Defects in this pathway, known as homologous recombination deficiency (HRD), can lead to genomic instability and are observed in approximately 13% of breast cancers. HRD is often driven by somatic or germline mutations in BRCA1/2, which render tumors sensitive to PARP inhibitors and platinum-based therapies through synthetic lethality. However, many HRD-positive cancers lack BRCA1/2 mutations, underscoring the need for more reliable approaches to identify tumors likely to benefit from these treatments.

## Pipeline

  To address this, we developed softHRD, a transcriptomics-based framework for detecting HRD in breast cancer. softHRD was trained on RNA-seq profiles from 857 breast cancer patients in The Cancer Genome Atlas (TCGA), filtered for protein-coding genes. A variational autoencoder was first used to reconstruct these transcriptomic profiles, generating latent representations that capture the underlying structure of gene expression patterns. A sparse autoencoder was then applied to these latent features to derive mechanistically interpretable components and identify an HRD-associated gene set. These genes were subsequently leveraged to train a downstream Elastic Net regression model, yielding a robust 111-gene transcriptional signature indicative of HRD. 

![pipeline](https://github.com/l1joseph/Multiscale_HRD_Classifier/blob/main/pipeline.png)

* ```./prelim_analysis/softHRD_pipeline.ipynb``` contains the full pipeline including external validation

### Deep learning based feature engineering 

![vae_sae](https://github.com/l1joseph/Multiscale_HRD_Classifier/blob/main/feat_eng.png)

![gene_imp](https://github.com/l1joseph/Multiscale_HRD_Classifier/blob/main/feat_map.png)

## Clinical validation

  We validated softHRD in 80 breast cancer patients from the I-SPY 2 clinical trial treated with neoadjuvant chemotherapy and olaparib. The model identified a statistically significant difference in pathologic complete response between HRD-predicted and HR-proficient tumors (p = 0.00676). Unlike whole-genome sequencing, which provides a static view of mutational alterations, transcriptomic profiling captures the dynamic state of gene expression, revealing biological changes that genomic methods may overlook.

![box](https://github.com/l1joseph/Multiscale_HRD_Classifier/blob/main/seaborn_ISPY_box.png)
![reg](https://github.com/l1joseph/Multiscale_HRD_Classifier/blob/main/seaborn_ISPY_reg.png)
  

## Benchmarking

  softHRD demonstrated robust performance across all PAM50 breast cancer subtypes, highlighting its generalizability. With the growing integration of transcriptomics into clinical research and diagnostics, softHRD represents a scalable and adaptable framework for accurate, efficient HRD characterization, with potential applications across multiple cancer types.
  
  ![bench](https://github.com/l1joseph/Multiscale_HRD_Classifier/blob/main/benchmark.png)
  



